### DISCUSSION TITLE
GitHub CoCalc usage

### DISCUSSION THREAD

**Discussion by nikkujensen** [sagemathinc/cocalc#6548] [General]
This may not be completely a CoCalc question, but here it is...

Can I use GitHub and CoCalc to clone and/or fork a repo to a folder in my CoCalc project? I don't want my whole CoCalc project stored/synced with GitHub. I want the GitHub repo to live in a folder in my CoCalc project.

**Comment by mforbes** [comment]
I typically clone all of my repositories into `~/.repositories` and then symlink them as needed.

Note: you have to be a bit careful when using `git` etc. on CoCalc since all the user is ~the project id~ `user`.  To deal with this, I forward the information when I SSH in.  See https://github.com/sagemathinc/cocalc/issues/370 for details.

If you will just use a terminal from the web interface, then you can define these variables in the terminal startup file (see https://doc.cocalc.com/terminal.html#startup-files for details.)

<details><summary>Details about what I actually do.</summary>

What I actually do is the following:

1. Create a new project and give it a license so I have internet access.
2. Setup an entry in my `~/.ssh/config` file on my laptop that looks like this:

    ```bash
    # ~/.ssh/config
    Host cc-project1
      User ff1cb986f...
    
    Host cc*
      HostName ssh.cocalc.com
      ForwardAgent yes
      SetEnv LC_HG_USERNAME=Your Full Name <your.email.address+hg@gmail.com>
      SetEnv LC_GIT_USERNAME=Your Full Name
      SetEnv LC_GIT_USEREMAIL=your.email.address+git@gmail.com
      SetEnv LC_EDITOR=vi
    ...
    ```

   Get the `User=` from your project Settings page (it is the project ID with no dashes).
3. Add my SSH keys to my Account page.
4. Login with SSH and run the following:

   ```bash
   ssh cc-project1
   pipx install mmf-setup
   mmf_setup cocalc
   ```

This installs a few tools I use (with `pipx`), then links some config files like `~/.bash_aliases` which then use these `LC_*` variables to tell `git` and `hg` who is logged in.  I have to use `LC_*` instead of directly setting `GIT_AUTHOR_NAME` etc. because SSH will not allow anything but `LC_*` variables to be sent.

Note: this is not an officially supported approach, but I use it daily and am reasonably happy with it.
</details>

**Reply by williamstein** [reply]
Note: On Cocalc.com the user is `user`. On cocalc-docker (and many years ago on cocalc.com), the user was the project_id with dashes removed.

**Reply by mforbes** [reply]
@williamstein Is there a better way of getting the project id than something like `$(shell uname -n | cut -c9-)`?

**Reply by williamstein** [reply]
Yes, the project id is $COCALC_PROJECT_ID.   That should work fine everywhere, even on cocalc.com, where the project id isn't easily related to the username.

Yes, we wrote the terminal startup files partly for exactly the same reason you're describing; often we have one file for each person, e.g., wstein.term.

**Reply by mforbes** [reply]
Where is  `$COCALC_PROJECT_ID` set?  I don't see it when I connect with `ssh`.  (It is there when I start terminal from `cocalc.com`.)  I might be mucking something up with my init files, but I don't see anything in the environment with `COCALC` in the name.

I think I asked this before, and I think the answer was no, but I don't recall why right now.  Is there any way that some environment variables could be set (`LC_HG_USERNAME` for my purposes) automatically when specific users open a new terminal from cocalc.com?  Something like the project-specific environment variables, but specified per user?  Of course, these would not work when connecting with `ssh`.  I think you mentioned possibly automatically generating user-specific terminals, but this misses (I think) the cases of people running commands in the Terminal window, or from notebooks, etc.

**Reply by mforbes** [reply]
One more question: It seems that the startup files execute, then the terminal is cleared.  Is there a way for the output of the startup file to be preserved so it is visible after it finishes executing?

**Reply by haraldschilly** [reply]
> when specific users open a new terminal

not for *new* terminals, but existing ones: the :rocket: icon creates a startup file associated with that terminal (and only the specific frame) https://doc.cocalc.com/terminal.html#startup-files

**Reply by haraldschilly** [reply]
ok, you know about startup files, about clearing them, I don't know. If I press Ctrl-D an output of `echo ...` shows up, but the first time around it is cleared. I'm guessing this could be changed.

**Reply by mforbes** [reply]
@haraldschilly I opened an issue https://github.com/sagemathinc/cocalc/issues/6551 with my use case.  *(Should I delete my conversation here about startup files to keep this thread cleaner and on-topic?)*

**Comment by williamstein** [comment]
> I want the GitHub repo to live in a folder in my CoCalc project.

I have a very different understanding of this question than @mforbes.  I think he's answering a pretty complicated question about how to host a git repo on CoCalc itself.       I suspect you're just asking about taking a repo from GitHub and storing a copy of that repo in CoCalc, then pushing/pull to/from GitHub.

To do that, just open a terminal via +New --> Terminal and use 100% the standard git commands.  There's nothing special at all about CoCalc compared to using Git (and GitHub) from any other Linux server, except be sure to setup your .gitconfig file with your email address (use `git config ...`). 

For example:

<img width="917" alt="image" src="https://user-images.githubusercontent.com/1276278/225199274-cce45ca9-7e78-4468-a551-f440e7dd5728.png">

**Reply by mforbes** [reply]
I am not really concerned with hosting the repo on CoCalc (though it is useful in some cases). My response is complicated because our projects have many users (a research group of students) who all commit, and we want to keep track of who commits: if there is only one user, then just adding the appropriate information with `git config ...` is the way to go.

**Comment by nikkujensen** [comment]
Thank you @williamstein for your [comment](https://github.com/sagemathinc/cocalc/discussions/6548#discussioncomment-5316752). This answers my inquiry

**Reply by williamstein** [reply]
I just added a chatgpt integration to the cocalc landing page, with some behind the scenes system prompting (and the chatgpt api).  Check out how it does on your question:

<img width="1303" alt="image" src="https://user-images.githubusercontent.com/1276278/225809660-d65e852d-7dc5-4603-9480-9afe7789c296.png">
